examples/cogvideo/train_cogvideox_lora.py

```diff
@@ -0,0 +1,1526 @@
+                    weight_dtype,
+                    requires_grad=args.train_text_encoder,
+                )
+
+                # Sample noise that will be added to the latents
+                noise = torch.randn_like(model_input)
+                batch_size, num_frames, num_channels, height, width = model_input.shape
+
+                # Sample a random timestep for each image
+                timesteps = torch.randint(
+                    0, scheduler.config.num_train_timesteps, (batch_size,), device=model_input.device
+                )
+                timesteps = timesteps.long()
+
+                # Prepare rotary embeds
+                image_rotary_emb = (
+                    prepare_rotary_positional_embeddings(
+                        height=args.height,
+                        width=args.width,
+                        num_frames=num_frames,
+                        vae_scale_factor_spatial=vae_scale_factor_spatial,
+                        patch_size=transformer.config.patch_size,
+                        attention_head_dim=transformer.config.attention_head_dim,
+                        device=accelerator.device,
+                    )
+                    if transformer.config.use_rotary_positional_embeddings
+                    else None
+                )
+
+                # Add noise to the model input according to the noise magnitude at each timestep
+                # (this is the forward diffusion process)
+                noisy_model_input = scheduler.add_noise(model_input, noise, timesteps)
+
+                # Predict the noise residual
+                model_output = transformer(
+                    hidden_states=noisy_model_input,
+                    encoder_hidden_states=prompt_embeds,
+                    timestep=timesteps,
+                    image_rotary_emb=image_rotary_emb,
+                    return_dict=False,
+                )[0]
+                model_pred = scheduler.get_velocity(model_output, noisy_model_input, timesteps)
+
+                alphas_cumprod = scheduler.alphas_cumprod[timesteps]
+                weights = 1 / (1 - alphas_cumprod)
+                while len(weights.shape) < len(model_pred.shape):
+                    weights = weights.unsqueeze(-1)
+
+                target = model_input
+
+                loss = torch.mean((weights * (model_pred - target) ** 2).reshape(batch_size, -1), dim=1)
+                loss = loss.mean()
+                accelerator.backward(loss)
+
+                if accelerator.sync_gradients:
+                    params_to_clip = (
+                        itertools.chain(transformer.parameters(), text_encoder.parameters())
+                        if args.train_text_encoder
+                        else transformer.parameters()
+                    )
+                    accelerator.clip_grad_norm_(params_to_clip, args.max_grad_norm)
+
+                optimizer.step()
+                lr_scheduler.step()
+                optimizer.zero_grad()
```

Comment:
again the optimzer should already been stepped at the `accelerator.backward(loss)` step when deepspeed is enabled
```suggestion
                if accelerator.state.deepspeed_plugin is None:
                    optimizer.step()
                    optimizer.zero_grad()
                lr_scheduler.step()
```

Reply:
We discussed internally and found that this is not needed and it will be a no-op based on different references that use accelerate

Reply:
the reason that's checked is because otherwise you have to create DummyOptim for deepspeed

Reply:
Based on [this](https://huggingface.co/docs/peft/v0.12.0/accelerate/deepspeed#the-important-parts), [this](https://github.com/huggingface/trl/blob/ebc85b2e39a046ea9d183720cb9b9396d61f2994/trl/trainer/ppo_trainer.py#L1100) and some other reference implementations, I think that it is not required. I was also able to train without this change.

I'm not too familiar with DeepSpeed and only recently started learning about it so if you've noticed different behaviour, happy to include these changes as they don't have any other side-effects I think.

Reply:
i'm not sure which config makes it manifest but i found it easiest to just create the dummyoptim for deepspeed. you're saying you have a real adamw instance and it works?

Reply:
I did have Adam, yes. This was my config:

```
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  gradient_accumulation_steps: 1
  gradient_clipping: 1.0
  offload_optimizer_device: none
  offload_param_device: none
  zero3_init_flag: false
  zero_stage: 2
distributed_type: DEEPSPEED
downcast_bf16: 'no'
enable_cpu_affinity: false
machine_rank: 0
main_training_function: main
mixed_precision: fp16
num_machines: 1
num_processes: 2
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

I have not exhaustively tested all DeepSpeed configurations but the current version of the script ran successfully on 2 A100s with a good final lora.

I'll read up on the DummyOptim part more and will add here most likely since I'm not too experienced with DeepSpeed, and might have gotten something wrong.

Reply:
I think I now see when this is needed. If DeepSpeed has an optimizer set, DummyOptim is required. In my case, I did not configure DeepSpeed to use an optimizer of its own.

https://huggingface.co/docs/accelerate/en/usage_guides/deepspeed

```
 # Creates Dummy Optimizer if `optimizer` was specified in the config file else creates Adam Optimizer
 optimizer_cls = (
     torch.optim.AdamW
     if accelerator.state.deepspeed_plugin is None
     or "optimizer" not in accelerator.state.deepspeed_plugin.deepspeed_config
     else DummyOptim
 )
 optimizer = optimizer_cls(optimizer_grouped_parameters, lr=args.learning_rate)
```